api/internal/localizer/localizer.go

```diff
@@ -0,0 +1,364 @@
+	}
+	dst := args.NewDir.Join(toDst)
+	if err = fSys.MkdirAll(dst); err != nil {
+		return errors.WrapPrefixf(err, "unable to create directory in localize destination")
+	}
+
+	depProvider := provider.NewDepProvider()
+	rFactory := resmap.NewFactory(depProvider.GetResourceFactory())
+	err = (&localizer{
+		fSys:      fSys,
+		ldr:       ldr,
+		dst:       filesys.ConfirmedDir(dst),
+		validator: depProvider.GetFieldValidator(),
+		rFactory:  rFactory,
+		pLdr:      pluginsLoader.NewLoader(krusty.MakeDefaultOptions().PluginConfig, rFactory, filesys.MakeFsOnDisk()),
+	}).localize()
+	if err != nil {
+		if errCleanup := fSys.RemoveAll(args.NewDir.String()); errCleanup != nil {
+			log.Printf("%s", errors.WrapPrefixf(errCleanup, "unable to clean localize destination").Error())
+		}
+		return errors.WrapPrefixf(err, "unable to localize target '%s'", targetArg)
+	}
+	return nil
+}
+
+func (lclzr *localizer) localize() error {
+	kt := target.NewKustTarget(lclzr.ldr, lclzr.validator, lclzr.rFactory, lclzr.pLdr)
+	err := kt.Load()
+	if err != nil {
+		return err
+	}
+
+	kust, err := lclzr.processKust(kt)
+	if err != nil {
+		return err
+	}
+
+	content, err := yaml.Marshal(kust)
+	if err != nil {
+		return errors.WrapPrefixf(
+			err,
+			"unable to serialize localized kustomization file for root '%s'", lclzr.ldr.Root())
+	}
+	return lclzr.fSys.WriteFile(lclzr.dst.Join(konfig.DefaultKustomizationFileName()), content)
+}
+
+func (lclzr *localizer) processKust(kt *target.KustTarget) (*types.Kustomization, error) {
+	kust := kt.Kustomization()
+
+	if path, exists := kust.OpenAPI["path"]; exists {
+		var err error
+		kust.OpenAPI["path"], err = lclzr.localizeFile(path)
+		if err != nil {
+			return &kust, err
+		}
+	}
+
+	for i, p := range kust.PatchesStrategicMerge {
+		if _, err := lclzr.rFactory.RF().SliceFromBytes([]byte(p)); err == nil { // inline
+			continue
+		}
+		newPath, err := lclzr.localizeFile(string(p)) // file
+		if err != nil {
+			return &kust, errors.WrapPrefixf(err, "unable to localize patchesStrategicMerge path")
+		}
+		kust.PatchesStrategicMerge[i] = types.PatchStrategicMerge(newPath)
+	}
+	for name, patches := range map[string][]types.Patch{
+		"patchesJson6902": kust.PatchesJson6902,
+		"patches":         kust.Patches,
+	} {
+		for i, p := range patches {
+			if p.Path == "" {
+				continue
+			}
+			newPath, err := lclzr.localizeFile(p.Path)
+			if err != nil {
+				return &kust, errors.WrapPrefixf(err, "unable to localize %s path", name)
+			}
+			patches[i].Path = newPath
+		}
+	}
+
+	for i, r := range kust.Replacements {
+		if r.Path == "" {
+			continue
+		}
+		newPath, err := lclzr.localizeFile(r.Path)
+		if err != nil {
+			return &kust, errors.WrapPrefixf(err, "unable to localize replacements path")
+		}
+		kust.Replacements[i].Path = newPath
+	}
+
+	for name, field := range map[string]*struct {
+		Mapper func(string) (string, error)
+		Paths  []string
+	}{
+		"resources": {
```

Comment:
The nice thing about the way you've done this is that it makes it really easy to handle other fields.

`bases` is deprecated, but we should be able to handle it exactly the same way as `resources`, so we could add that here as well.

Reply:
I don't think I'll encounter `bases` because [they were moved into `resources` when the kustomization was first loaded](https://github.com/kubernetes-sigs/kustomize/blob/master/api/types/kustomization.go#L190-L191).